pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -239,6 +239,21 @@ private String getQueryResponse(String query, @Nullable SqlNode sqlNode, String
       tableName = _pinotHelixResourceManager.getActualTableName(inputTableName);
     } catch (Exception e) {
       LOGGER.error("Caught exception while compiling query: {}", query, e);
+      try {
+        // try to compile the query using multi-stage engine and suggest using it if it succeeds.
+        LOGGER.info("Trying to compile query {} using multi-stage engine", query);
+        QueryEnvironment queryEnvironment = new QueryEnvironment(new TypeFactory(new TypeSystem()),
+            CalciteSchemaBuilder.asRootSchema(new PinotCatalog(_pinotHelixResourceManager.getTableCache())), null,
+            null);
+        queryEnvironment.getTableNamesForQuery(query);
+        LOGGER.info("Successfully compiled query using multi-stage engine: {}", query);
+        return QueryException.getException(QueryException.SQL_PARSING_ERROR, new Exception(
+            "It seems that the query contains join, please try the multi-stage engine by checking the "
```

Comment:
The query may contain other multi-stage execution plan, e.g. (joins, sub-query, set operation, etc)

Reply:
good point, will update the error message